staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/stdlib.go

```diff
@@ -0,0 +1,616 @@
+		}
+	}
+	return types.Int(-1)
+}
+
+func lastIndexOf(list ref.Val, item ref.Val) ref.Val {
+	lister, ok := list.(traits.Lister)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	sz := lister.Size().Value().(int64)
+	for i := sz - 1; i >= 0; i-- {
+		if lister.Get(types.Int(i)).Equal(item) == types.True {
+			return types.Int(i)
+		}
+	}
+	return types.Int(-1)
+}
+
+func find(strVal ref.Val, regexVal ref.Val) ref.Val {
+	str, ok := strVal.Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	regex, ok := regexVal.Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	re, err := regexp.Compile(regex)
+	if err != nil {
+		return types.NewErr("Illegal regex: %v", err.Error())
+	}
+	result := re.FindString(str)
+	return types.String(result)
+}
+
+func findAll(args ...ref.Val) ref.Val {
+	argn := len(args)
+	if argn < 2 || argn > 3 {
+		return types.NoSuchOverloadErr()
+	}
+	str, ok := args[0].Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	regex, ok := args[1].Value().(string)
+	if !ok {
+		return types.NoSuchOverloadErr()
+	}
+	n := int64(-1)
+	if argn == 3 {
+		n, ok = args[2].Value().(int64)
+		if !ok {
+			return types.NoSuchOverloadErr()
+		}
+	}
+
+	re, err := regexp.Compile(regex)
+	if err != nil {
+		return types.NewErr("Illegal regex: %v", err.Error())
+	}
+
+	result := re.FindAllString(str, int(n))
+
+	return types.NewStringList(types.DefaultTypeAdapter, result)
+}
+
+func stringToUrl(arg ref.Val) ref.Val {
+	s := arg.Value().(string)
+	u, err := url.Parse(s)
+	if err != nil {
+		return types.NewErr("URL parse error during conversion from string: %v", err)
+	}
+	return model.URL{URL: u}
+}
+
+func urlToString(arg ref.Val) ref.Val {
+	s := arg.Value().(*url.URL)
+	return types.String(s.String())
+}
+
+func getScheme(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+	return types.String(u.Scheme)
+}
+
+func getHostname(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+	return types.String(u.Hostname())
+}
+
+func getPort(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+	return types.String(u.Port())
+}
+
+func getUsername(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+	return types.String(u.User.Username())
+}
+
+func getPassword(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+	p, ok := u.User.Password()
+	if !ok {
+		return types.String("")
+	}
+	return types.String(p)
+}
+
+func getPath(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+	return types.String(u.Path)
+}
+
+func getQuery(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+
+	result := map[ref.Val]ref.Val{}
+	for k, v := range u.Query() {
+		result[types.String(k)] = types.NewStringList(types.DefaultTypeAdapter, v)
+	}
+	return types.NewRefValMap(types.DefaultTypeAdapter, result)
+}
+
+func getFragment(arg ref.Val) ref.Val {
+	u := arg.Value().(*url.URL)
+	return types.String(u.EscapedFragment())
+}
+
+func isURL(arg ref.Val) ref.Val {
+	s := arg.Value().(string)
+	_, err := url.Parse(s)
```

Comment:
same comment about url.Parse being very lenient about what it accepts; is the expectation this only accepts absolute urls that specify scheme and host?

Reply:
ParseRequestURI guard added